Common/Data/DividendYieldProvider.cs

```diff
@@ -0,0 +1,188 @@
+ * limitations under the License.
+*/
+
+using QuantConnect.Configuration;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Util;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+
+namespace QuantConnect.Data
+{
+    /// <summary>
+    /// Estimated annualized continuous dividend yield at given date
+    /// </summary>
+    public class DividendYieldProvider : IDividendYieldModel
+    {
+        private Symbol _symbol;
+        private static DateTime _firstDividendYieldDate = new DateTime(1998, 1, 1);
+        private static DateTime _lastDividendYieldDate;
+        private static Dictionary<DateTime, decimal> _dividendYieldRateProvider;
+        private static readonly object _lock = new();
+
+        /// <summary>
+        /// Default no dividend payout
+        /// </summary>
+        public static readonly decimal DefaultDividendYieldRate = 0.0m;
+
+        /// <summary>
+        /// Instantiates a <see cref="DividendYieldProvider"/> with the specified Symbol
+        /// </summary>
+        public DividendYieldProvider(Symbol symbol)
+        {
+            _symbol = symbol;
+        }
+
+        /// <summary>
+        /// Lazily loads the dividend provider from disk and returns it
+        /// </summary>
+        private IReadOnlyDictionary<DateTime, decimal> DividendYieldRateProvider
+        {
+            get
+            {
+                // let's not lock if the provider is already loaded
+                if (_dividendYieldRateProvider != null)
+                {
+                    return _dividendYieldRateProvider;
+                }
+
+                lock (_lock)
+                {
+                    if (_dividendYieldRateProvider == null)
+                    {
+                        LoadDividendYieldProvider();
+                    }
+                    return _dividendYieldRateProvider;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get dividend yield by a given date
+        /// </summary>
+        /// <param name="date">The date</param>
+        /// <returns>Dividend yield on the given date</returns>
+        public decimal GetDividendYield(DateTime date)
+        {
+            if (!DividendYieldRateProvider.TryGetValue(date.Date, out var dividendYield))
+            {
+                return date < _firstDividendYieldDate
+                    ? DefaultDividendYieldRate
+                    : DividendYieldRateProvider[_lastDividendYieldDate];
+            }
+
+            return dividendYield;
+        }
+
+        /// <summary>
+        /// Generate the daily historical dividend yield
+        /// </summary>
+        protected void LoadDividendYieldProvider()
+        {
+            var directory = Path.Combine(Globals.DataFolder, "equity", "usa", "factor_files", $"{_symbol.Value.ToLowerInvariant()}.csv");
+            _dividendYieldRateProvider = FromCsvFile(directory);
```

Comment:
-  Don't think we should directly access the csv, they might not be there when using the zip factor file provider, they can be in cache etc. Believe we could do something like the following:

```csharp
            var factorFileProvider = Composer.Instance.GetPart<IFactorFileProvider>();

            foreach (var factorRow in factorFileProvider.Get(_symbol))
            {
                var corporateFactor = factorRow as CorporateFactorRow;
                if (corporateFactor != null)
                {

                }
            }
```


- Also note that for future options, index options, the dividend yield provider shouldn't be really used?
- For live trading we will need to refresh the dividend yield, thinking about the long lived deployment case. Please take a look at `LocalZipFactorFileProvider` and how it solves it, think we can copy it over here 👍

Reply:
The last live trading point needs to be addressed 👍